src/Columns/ColumnDynamic.cpp

```diff
@@ -624,27 +843,142 @@ int ColumnDynamic::doCompareAt(size_t n, size_t m, const DB::IColumn & rhs, int
     else if (right_discr == ColumnVariant::NULL_DISCRIMINATOR)
         return -nan_direction_hint;
 
-    /// If rows have different types, we compare type names.
-    if (variant_info.variant_names[left_discr] != right_dynamic.variant_info.variant_names[right_discr])
-        return variant_info.variant_names[left_discr] < right_dynamic.variant_info.variant_names[right_discr] ? -1 : 1;
+    /// Check if both values are in shared variant.
+    if (left_discr == left_shared_variant_discr && right_discr == right_shared_variant_discr)
+    {
+        /// Extract type names from both values.
```

Comment:
can you have a fast path for equal values if you just compare `left_value` with `right_value`?
or can the types we treat as equal have different serialization?

Reply:
It's a good idea! No, equal types should not have different serialization.